### PR TITLE
Element vs Section fix (issue #213)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review.

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -3954,18 +3954,18 @@ class Variable_Name(Base):  # R602
...
+    Fortran 2003 rule 603
+
+    designator is object-name
+               or array-element
+               or array-section
+               or structure-component
+               or substring
+
+    '''
+    subclass_names = ['Object_Name', 'Array_Element', 'Array_Section',
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py]: Please add a comment here explaining that we are going to fail to match Substring with a reference to the relevant Issue (and the Fortran standard).

File: src/fparser/two/tests/fortran2003/test_designator.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+'''
+import pytest
+from fparser.two.Fortran2003 import Designator, Name, Part_Ref, \
+    Array_Section, Program, Substring, Data_Ref
+from fparser.api import get_reader
+
+
+@pytest.mark.usefixtures("f2003_create")
+def test_object_name():
+    '''Test that an Name object is returned when a valid object-name name
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_designator.py]: oops.

File: src/fparser/two/tests/fortran2003/test_designator.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    '''
+    result = Designator("fred")
+    assert str(result) == "fred"
+    assert isinstance(result, Name)
+
+
+@pytest.mark.usefixtures("f2003_create")
+def test_array_element():
+    '''Test that a Part_Ref object is returned when a valid array-element
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_designator.py]: Actually the rule was meant to test the array_element rule but I had messed up. I've now changed the rule and added some extra words about rules wrongly matching. Hopefully this makes it clearer.

File: src/fparser/two/tests/fortran2003/test_designator.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    There are various constraints in Array_Section which might
+    distinguish between the rules but I wonder whether a different
+    ordering would be required as I am not sure that the constraints
+    on earlier rules avoid overlaps with later rules.
+
+    '''
+    reader = get_reader(
+        "program test\n"
+        "character(len=10) :: a\n"
+        "a(1:3)='hey'\n"
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_designator.py]: The explanation is that I messed up :-). I've hopefully fixed the tests now.

File: src/fparser/two/tests/fortran2003/test_designator.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+
+@pytest.mark.usefixtures("f2003_create")
+def test_structure_component():
+    '''Test that a Structure_Component object is returned when a valid
+    structure-component is supplied.
+
+    *** This rule is never called as it has the same structure as
+    Array_Element which precedes it in the list and is therefore
+    matched first ***
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_designator.py]: Hopefully the text now explains.

File: src/fparser/two/tests/fortran2003/test_designator.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    matched first. However, this in turn is not called as it is
+    preceded by Array_Element which is matched first. ***
+
+    There are various constraints in Substring which might distinguish
+    between the rules but I wonder whether a different ordering would
+    be required as I am not sure that the constraints on earlier rules
+    avoid overlap with later rules.
+
+    '''
+    result = Designator("parent_string(1:2)")
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_designator.py]: I agree. There is nothing that can be done here as there is no context information. I've dropped this test as a result and have a test with contextual information instead.

File: src/fparser/two/tests/fortran2003/test_designator.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    assert str(result) == "a(:)(:)"
+    assert isinstance(result, Array_Section)
+
+
+@pytest.mark.xfail(reason="issue #201 Array section is parsed as an array "
+                   "element.")
+@pytest.mark.usefixtures("f2003_create")
+def test_array_section2():
+    '''Test that an Array_Section object is returned when a valid
+    array-element is supplied. This fails as the example matches with
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_designator.py]: I've restructured the tests so hopefully they are correct now.

File: src/fparser/two/Fortran2003.py (hunk 1 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Good idea. Done.

**Comment by rupertford** [comment]
I think we're ready for the next review.

**Comment by rupertford** [comment]
Ready for next review subject to travis being happy.

**Comment by arporter** [comment]
Documentation is much improved - thanks. I've taken the liberty of fixing a few typos.

**Pull request merged**